soh/soh/Enhancements/item-tables/ItemTableManager.cpp

```diff
@@ -0,0 +1,59 @@
+#include "ItemTableManager.h"
+
+ItemTableManager::ItemTableManager() {
+}
+
+ItemTableManager::~ItemTableManager() {
+    this->itemTables.clear();
+}
+
+bool ItemTableManager::AddItemTable(uint16_t tableID) {
+    ItemTable newItemTable;
+    return itemTables.emplace(tableID, newItemTable).second;
+}
+
+bool ItemTableManager::AddItemEntry(uint16_t tableID, uint16_t getItemID, uint16_t itemID, uint16_t objectID, int16_t drawID, uint16_t textID, uint16_t field, bool chestAnim) {
+    ItemTable* itemTable = RetrieveItemTable(tableID);
+    if (itemTable == nullptr) {
+        return false;
+    }
+    GetItemEntry getItemEntry = GET_ITEM(itemID, objectID, drawID, textID, field, chestAnim);
+    return itemTable->emplace(getItemID, getItemEntry).second;
+}
+
+bool ItemTableManager::AddItemEntry(uint16_t tableID, uint16_t getItemID, GetItemEntry getItemEntry) {
+    ItemTable* itemTable = RetrieveItemTable(tableID);
+    if (itemTable == nullptr) {
+        return false;
+    }
+    return itemTable->emplace(getItemID, getItemEntry).second;
+}
+
+GetItemEntry ItemTableManager::RetrieveItemEntry(uint16_t tableID, uint16_t itemID) {
+    ItemTable* itemTable = RetrieveItemTable(tableID);
+    if (itemTable != nullptr) {
+        auto foundItemEntry = itemTable->find(itemID);
+        if (foundItemEntry != itemTable->end()) {
+            return foundItemEntry->second;
```

Comment:
Well the idea is that the `foundItemEntry` is used to verify whether or not we were able to find the `itemEntry` and by the time we assign anything to the `itemEntry` variable it is just the GetItemEntry and not the whole pair. In hindsight, it might be more readable to use `unordered_map::at` to get the itemEntry, and return GET_ITEM_NONE when the not found exception is caught instead.

Reply:
And this would apply to custom_messages as well so I'd probably want to update that PR with the same thing if that sounds better to you.

Reply:
hmm, i don't really have strong feelings one way or the other about the current method vs a try/catch. if the try/catch feels like it'll be more readable let's go for it! (it'll be fine to update the message stuff to use that pattern in this PR)